### ISSUE TITLE
Multiple columns key

### ISSUE THREAD

**Issue by vlad-bezden** [datafold/data-diff#110]
I have a database with multiple columns as a primary key. Is there any way to specify multiple columns as a key?

**Comment by erezsh** [comment]
This feature hasn't been implemented yet, but it's definitely something we are going to support.

**Comment by yuzeh** [comment]
Is there a place where we can register our support for this feature? Appreciate it!

**Comment by erezsh** [comment]
@yuzeh Here is a good start :)

**Comment by sirupsen** [comment]
@yuzeh @vlad-bezden two questions (and for anyone else weighing in wanting this):

1. What database(s) are you using, and do they both use a multi-column primary key?
2. What column types are in the primary key?

**Comment by avaitla** [comment]
I use mysql and have data schemas with composite primary keys as well:

```
CREATE TABLE `payments` (
  `customerNumber` int NOT NULL,
  `checkNumber` varchar(50) NOT NULL,
  `paymentDate` date NOT NULL,
  `amount` decimal(10,2) NOT NULL,
  PRIMARY KEY (`customerNumber`,`checkNumber`),
  CONSTRAINT `payments_ibfk_1` FOREIGN KEY (`customerNumber`) REFERENCES `customers` (`customerNumber`)
) ENGINE=InnoDB DEFAULT CHARSET=latin1;
```

Taken from https://www.mysqltutorial.org/mysql-sample-database.aspx

**Comment by erezsh** [comment]
@avaitla @vlad-bezden @yuzeh The PR for a compound primary key is ready!

You can install it with:

```shell
pip install git+https://github.com/datafold/data-diff.git@refs/pull/158/head
```
And then you can use multiple `-k` arguments, for example:

```shell
data_diff postgresql:/// table1 postgresql:/// table2 -k id -k id2 -t timestamp
```

**Comment by erezsh** [comment]
@vlad-bezden @yuzeh @avaitla I had some time to think about this problem, and I think it should work with the current version (not the new PR), by setting `-k` to the column with the most distinct values. You can include the other primary key columns with `-c`.

I don't think it will be easy to improve on this method, in terms of performance, and it seems to be the simplest way forward.

Can you give it a try?

If that's doesn't answer your situation, please let me know.

**Comment by yuzeh** [comment]
Perfect. Test driving now.

**Comment by yuzeh** [comment]
@erezsh Did the test, unfortunately Got this error:

```
NotImplementedError: Cannot use a column of type Text() as a key
```

Any thoughts on how to work around?

**Comment by yuzeh** [comment]
@sirupsen Missed your question here, apologies!

> @yuzeh @vlad-bezden two questions (and for anyone else weighing in wanting this):
> 
> 1. What database(s) are you using, and do they both use a multi-column primary key?
> 2. What column types are in the primary key?

1. BigQuery. There's no concept of a primary key in BQ, the closest you could get is to designate a combination of columns as "clustering columns"
2. string, timestamp, date

**Comment by erezsh** [comment]
@yuzeh Can you please share the type of the column you use as a key, and a few sample values from that column?

**Comment by GCCree** [comment]
@erezsh Where I work we're suffering from the compound key issue, and we're floating the idea of trying to extend the data-diff library to support compound keys ourselves. I'm just curious, how far along is this feature?

**Comment by erezsh** [comment]
@GCCree I am close to finishing a new implementation of compound keys. I will probably post the PR next week.

**Comment by erezsh** [comment]
Just FYI for anyone that's interested, the PR is already up: https://github.com/datafold/data-diff/pull/375

If you want to help us out, take it for a spin and let us know how well it worked!

I hope we'll merge it soon, once the code review process is done.
